### ISSUE TITLE
emplace fails to supply a shared_service

### ISSUE THREAD

**Issue by giacomocariello** [gracicot/kangaru#59]
If I use a shared_service with kgr::supplied to pass some arbitrary values to constructor, when I try to emplace(), I get a compilation error. I thought that shared_service, by inheriting kgr::single, could be used with kgr::supplied.

**Comment by gracicot** [comment]
Minimal code to reproduce:
```c++
struct Test { Test(int) {} };
struct TestService : kgr::shared_service<Test> {};

int main() {
    kgr::container{}.emplace<TestService>(1);
}
```

**Comment by giacomocariello** [comment]
Yep. For completeness, I actually added kgr::supplied as base class for TestService, but the error output is identical.

error: no matching function for call to 'kgr::container::emplace(int)'
         kgr::container{}.emplace<TestService>(1);
                                                ^
In file included from kangaru/include/kangaru/autocall.hpp:9:0,
                 from kangaru/include/kangaru/kangaru.hpp:4,
                 from main.cpp:8:
kangaru/include/kangaru/container.hpp:95:7: note: candidate: template<class T, class ... Args, typename std::enable_if<std::is_base_of<kgr::single, T>::value, int>::type <anonymous>, typename std::enable_if<kgr::detail::is_construction_valid<T, Args ...>::value, int>::type <anonymous> > bool kgr::container::emplace(Args&& ...)
  bool emplace(Args&&... args) {
       ^
kangaru/include/kangaru/container.hpp:95:7: note:   template argument deduction/substitution failed:
kangaru/include/kangaru/container.hpp:94:58: error: no type named 'type' in 'struct std::enable_if<false, int>'
   enable_if<detail::is_construction_valid<T, Args...>> = 0>
                                                          ^
kangaru/include/kangaru/container.hpp:94:58: note: invalid template non-type parameter
kangaru/include/kangaru/container.hpp:106:7: note: candidate: template<class T, typename std::enable_if<std::is_default_constructible<kgr::detail::service_error<T> >::value, int>::type <anonymous> > bool kgr::container::emplace(kgr::detail::service_error<T>) <deleted>
  bool emplace(detail::service_error<T> = {}) = delete;
       ^
kangaru/include/kangaru/container.hpp:106:7: note:   template argument deduction/substitution failed:
main.cpp:22:48: note:   cannot convert '1' (type 'int') to type 'kgr::detail::service_error<TestService>'
         kgr::container{}.emplace<TestService>(1);
                                                ^
...(omissis)...

**Comment by gracicot** [comment]
Thank you for your report. It was an oversight from my part when adding the emplace function into the container.

Before version 4, single services were not constructible with arguments. So thier construct function were not parameterized. If you look at [this line](https://github.com/gracicot/kangaru/blob/c91813896bec0b9b8ee01fda1b960ff95a8d5302/include/kangaru/service.hpp#L143), you'll see that the construct function only take the dependencies as argument, whereas all other generic service accept additional arguments.

Note that a single service don't have to inherit from `kgr::supplied` to be useable with emplace, they only need to be single. `kgr::supplied` only tell the container to not try to construct that service automatically as it may need additional arguments at construction, like my `Test` class above.

I got a local patch for that, and I'm currently writing unit tests for that. Once working on all platform, I'll update this issue. Expect the fix to be in `v4.0.1`

**Comment by giacomocariello** [comment]
Great, thanks!

**Comment by gracicot** [comment]
The patch passes the tests on all compilers. It has been fixed in [416962e](https://github.com/gracicot/kangaru/commit/416962e68f36ebca9772d4abf2a397878224d2a4) and merged to master.

If there's anything else, feel free to open other issues. It's always a pleasure to help!
